File: Crawler/Output/OutputType.php
<?php

namespace Crawler\Output;

enum OutputType implements \JsonSerializable
{
    case FORMATTED_TEXT;
    case JSON;
    case MULTI;

    public static function fromText(string $text): self
    {
        $text = trim(strtolower($text));
        if ($text === 'text') {
            return self::FORMATTED_TEXT;
        } elseif ($text === 'json') {
            return self::JSON;
        } else {
            throw new \Exception("Unknown output type '{$text}'. Supported values are: " . implode(', ', self::getAvailableTextTypes()));
        }
    }

    /**
     * @return string[]
     */
    public static function getAvailableTextTypes(): array
    {
        return ['text', 'json'];
    }

    public function jsonSerialize(): mixed
    {
        if ($this === self::FORMATTED_TEXT) {
            return 'text';
        } elseif ($this === self::JSON) {
            return 'json';
        } elseif ($this === self::MULTI) {
            return 'multi';
        } else {
            throw new \Exception("Unknown output type '{$this}'");
        }
    }
}

File: src/output/output_type.rs
// SiteOne Crawler - OutputType enum
// (c) Jan Reges <jan.reges@siteone.cz>

use serde::{Deserialize, Serialize};
use std::fmt;

use crate::error::CrawlerError;

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OutputType {
    Text,
    Json,
    Multi,
}

impl OutputType {
    pub fn from_text(text: &str) -> Result<Self, CrawlerError> {
        match text.trim().to_lowercase().as_str() {
            "text" => Ok(OutputType::Text),
            "json" => Ok(OutputType::Json),
            other => Err(CrawlerError::Parse(format!(
                "Unknown output type '{}'. Supported values are: {}",
                other,
                Self::available_text_types().join(", ")
            ))),
        }
    }

    pub fn available_text_types() -> Vec<&'static str> {
        vec!["text", "json"]
    }

    pub fn as_str(&self) -> &'static str {
        match self {
            OutputType::Text => "text",
            OutputType::Json => "json",
            OutputType::Multi => "multi",
        }
    }
}

impl fmt::Display for OutputType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(self.as_str())
    }
}
